src/serde/ser_br.rs

```diff
@@ -0,0 +1,84 @@
+// Serialization with "back-references"
+
+use std::io;
+use std::io::Cursor;
+
+use super::object_cache::{serialized_length, treehash, ObjectCache};
+use super::read_cache_lookup::ReadCacheLookup;
+use super::write_atom::write_atom;
+use crate::allocator::{NodePtr, SExp};
+use crate::node::Node;
+
+const BACK_REFERENCE: u8 = 0xfe;
+const CONS_BOX_MARKER: u8 = 0xff;
+
+#[derive(PartialEq, Eq)]
+enum ReadOp {
+    Parse,
+    Cons,
+}
+
+// these test cases were produced by:
+
+// from chia.types.blockchain_format.program import Program
+// a = Program.to(...)
+// print(bytes(a).hex())
+// print(a.get_tree_hash().hex())
+
+pub fn node_to_stream_backrefs(node: &Node, f: &mut dyn io::Write) -> io::Result<()> {
```

Comment:
I think you should make `f` a template instead of using dynamic dispatch. I think you can do that like this:
```
pub fn<S: io::Write> node_to_stream_backrefs(node: &Node, f: &mut S) -> io::Result<()> {
```

Reply:
I proposed this, related, change: https://github.com/Chia-Network/clvm_rs/pull/221

Reply:
This has been turned into a template, so it's the best of both worlds (at the cost of possibly slower compilation).